File: src/main/java/com/baidu/algorithm/dfs/_140_Word_Break_II.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: List[str]
        """
        if not s:
            return []

        return self.dfs(s, wordDict, {})

    def dfs(self, s, words, mp):
        if s in mp:
            return mp[s]

        if not s:
            return [""]

        res = []
        for word in words:
            if s.startswith(word):
                sub_res = self.dfs(s[len(word):], words, mp)
                for x in sub_res:
                    res.append(word + ("" if not x else " ") + x)

        mp[s] = res
        return res

File: src/main/java/com/baidu/algorithm/dfs/_140_Word_Break_II.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.dfs;

import java.util.*;

/**
 * _140_Word_Break_II
 *
 * @author xuhaoran01
 */
public class _140_Word_Break_II {

    private List<String> dfs(String s, List<String> wordDict, Map<String, List<String>> cache) {

        if (cache.containsKey(s)) {
            return cache.get(s);
        }

        if (s.isEmpty()) {
            return new ArrayList<>(Arrays.asList(""));
        }

        List<String> res = new ArrayList<>();
        for (String word : wordDict) {
            if (s.startsWith(word)) {
                List<String> subRes = dfs(s.substring(word.length()), wordDict, cache);
                for (String sub : subRes) {
                    res.add(sub.isEmpty() ? word : word + " " + sub);
                }
            }
        }

        cache.put(s, res);

        return res;
    }

    public List<String> wordBreak(String s, List<String> wordDict) {
        List<String> res = new ArrayList<>();
        if (s == null || s.isEmpty()) {
            return res;
        }

        return dfs(s, wordDict, new HashMap<>());
    }
}
